lib/model/avatar_url.dart

```diff
@@ -0,0 +1,148 @@
+import 'dart:convert';
+
+import 'package:crypto/crypto.dart';
+import 'package:path/path.dart';
+
+import '../log.dart';
+
+/// Pixel dimensions of different size choices we have (they're all
+/// square) when requesting an uploaded avatar.
+/// DEFAULT_AVATAR_SIZE in zerver/lib/upload.py.
+const int defaultUploadSizePx = 100;
+
+/// MEDIUM_AVATAR_SIZE in zerver/lib/upload.py.
+const int mediumUploadSizePx = 500;
+
+abstract class AvatarURL {
+  static AvatarURL fromUserOrBotData({
+    String? rawAvatarUrl,
+    required Uri realm,
+    required int userId,
+    required String email,
+  }) {
+    if (rawAvatarUrl == null || rawAvatarUrl.isEmpty) {
+      return GravatarURL.validateAndConstructInstance(email: email);
+    } else if (rawAvatarUrl.startsWith(GravatarURL.origin)) {
+      return GravatarURL.validateAndConstructInstance(
+          urlFromServer: rawAvatarUrl, email: email);
+    } else if (['.png','.jpg','.jpeg'].contains(extension(rawAvatarUrl))) {
+      return UploadedAvatarURL.validateAndConstructInstance(
+          realm: realm, absoluteOrRelativeUrl: rawAvatarUrl);
+    } else if (rawAvatarUrl.isNotEmpty) {
+      return FallbackAvatarURL.validateAndConstructInstance(
+          realm: realm, userId: userId);
+    }
+
+    const msg =
+        'Unexpected value for `rawAvatarUrl` in `AvatarURL.fromUserOrBotData`';
+    assert(debugLog(('$msg value:$rawAvatarUrl')));
+    throw Error();
+  }
+
+  Uri get(double sizePhysicalPx);
+}
+
+class GravatarURL implements AvatarURL {
+  static GravatarURL validateAndConstructInstance({
+    required String email,
+    String? urlFromServer,
+  }) {
+    if (urlFromServer != null) {
+      return GravatarURL(standardUri: Uri.parse(urlFromServer));
+    } else {
+      return GravatarURL(
+          standardUri: Uri.parse(
+              '${GravatarURL.origin}/avatar/${md5.convert(utf8.encode(email.toLowerCase()))}?d=identicon'));
+    }
+  }
+
+  static String origin = 'https://secure.gravatar.com';
+
+  Uri standardUri;
+
+  GravatarURL({required this.standardUri});
+
+  @override
+  Uri get(double sizePhysicalPx) {
+    Uri result = standardUri.replace(queryParameters: {
+      ...standardUri.queryParameters,
+      's': sizePhysicalPx.toString(),
+    });
+    return result;
+  }
+}
+
+class FallbackAvatarURL extends AvatarURL {
+  static FallbackAvatarURL validateAndConstructInstance(
+      {required Uri realm, required int userId}) {
+    return FallbackAvatarURL(
+        standardUri: Uri.parse('${realm.origin}/avatar/${userId.toString()}'));
+  }
+
+  Uri standardUri;
+
+  FallbackAvatarURL({required this.standardUri});
+
+  @override
+  Uri get(double sizePhysicalPx) {
+    Uri result = standardUri;
+    if (sizePhysicalPx > defaultUploadSizePx) {
+      result = Uri.parse('$standardUri/medium');
+    }
+
+    return result;
+  }
+}
+
+class UploadedAvatarURL extends AvatarURL {
+  static UploadedAvatarURL validateAndConstructInstance({
```

Comment:
The Dart idiom here is to make this a factory constructor:
```suggestion
  factory UploadedAvatarURL({
```

And then the other constructor below can be private:
```dart
  UploadedAvatarURL._({required this.standardUri});
```

See our `Unreads` for an example. Or find other examples with: `git grep -P '\w+\._\('`

Reply:
[d1cd89f](https://github.com/zulip/zulip-flutter/pull/534/commits/d1cd89f04490ac4053ab7ce995d83baad1656f3f)